diff --git a/ast3/Include/token.h b/ast3/Include/token.h
index a657fdd..d0b2b94 100644
--- a/ast3/Include/token.h
+++ b/ast3/Include/token.h
@@ -68,8 +68,10 @@ extern "C" {
 /* These aren't used by the C tokenizer but are needed for tokenize.py */
 #define COMMENT         55
 #define NL              56
-#define ENCODING                57
-#define N_TOKENS        58
+#define ENCODING        57
+#define TYPE_IGNORE     58
+#define TYPE_COMMENT    59
+#define N_TOKENS        60
 
 /* Special definitions for cooperation with parser */
 
